src/hub.js

```diff
@@ -117,79 +106,122 @@ export function hub(chrome) {
+			appSpawns.push({ appId, message });
+		} catch (e) {
+			postMessage({
+				type: 'SPAWN-FAIL',
+				source: CHROME_APP_ID,
+				target: message.source,
+				token: appWindows.get(windowByApp(CHROME_APP_ID)).token,
+				topic: message.topic,
+				data: message.target + " is not activated on the current user's account"
+			});
+		}
+	}
+
+	function handlePong(event) {
+		const token = event.data.token;
+
+		appPongs.set(token, {
+			...appPongs.get(token),
+			lastPong: window.performance.now()
+		});
+	}
+
+	function handleEvent({ data: message, source: sourceWindow }) {
+		/**
+		 * @TODO Handle the case when the Chrome blocks certain targets for certain sources
+		 */
+		const targetWindow = windowByApp(message.target, sourceWindow);
+		debug(
+			'Routing %o from %o to %o - %O',
+			message.type,
+			message.source,
+			message.target,
+			message
+		);
+		postMessage(message, targetWindow);
+
+		if (message.type.indexOf('SPAWN') === 0) {
+			handleAppSubmit(message.source, message.target, message.data);
+		}
+	}
+
 	window.addEventListener('message', function(event) {
 		const message = event.data;
+		const appWindow = appWindows.get(event.source) || {};
+		message.source = appWindow.appId;
+		message.viaHub = true;
 
 		// Only accept valid messages from apps.
 		if (!hubMessageValid(message)) {
 			return;
 		}
 
 		// Message from a frame we don't know yet.
-		if (!appWindows.has(event.source)) {
-			// The only command should be CONNECT, we fail otherwise.
-			if (message.type === 'CONNECT') {
-				const appId = appByWindow(event.source);
-				const token = uuid();
-				debug('CONNECT %o', appId);
-				appWindows.set(event.source, { appId, token });
-				const spawnWaiting = appSpawns.findIndex(
-					spawn => spawn.appId === appId
-				);
-				const connackMessage = {
-					type: 'CONNACK',
-					viaHub: true,
-					target: appId,
-					token
-				};
-				if (spawnWaiting !== -1) {
-					const { data, source } = appSpawns[spawnWaiting].message;
-					connackMessage.source = source;
-					connackMessage.data = data;
-					appSpawns.splice(spawnWaiting, 1);
-				}
-				postMessage(connackMessage, event.source);
-				const pingOpts = { targetWindow: event.source, appId, token };
-				const timeoutId = setTimeout(
-					() => pingApp(pingOpts),
-					hubInstance.HEARTBEAT
-				);
-				appPongs.set(token, {
-					lastPong: window.performance.now(),
-					timeoutIds: new Set([timeoutId])
-				});
-				return;
-			} else {
-				console.warn(
-					'Unexpected critical error! ts.app sent message without being connected!',
-					event
-				);
-				return;
-			}
-		}
-
-		if (message.token !== appWindows.get(event.source).token) {
+		// The only command should be CONNECT, we fail otherwise.
+		if (!appWindow && message.type !== 'CONNECT') {
+			console.warn(
+				'Unexpected critical error! app sent message without being connected!',
+				event
+			);
+			return;
+		} else if (message.token !== appWindow.token) {
```

Comment:
tricky ;)

for `CONNECT` this will be `undefined !== undefined` , that should be ok